Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

```diff
@@ -0,0 +1,573 @@
+`(h.codChart.extend J).target = (h.codChart.extend I) '' h.codChart.source`,
+but that does *not* imply `f` maps `h.domChart.source` to `h.codChartSource`;
+a priori `f` could map some point `f ∘ h.domChart.extend I x ∉ h.codChart.source` into the target.
+Note that this difference only occurs because of our design using junk values;
+this is not a mathematically meaningful difference.`
+
+At the same time, this condition is fairly weak: it is implied, for instance, by `f` being
+continuous at `x` (see `mk_of_continuousAt`), which is easy to ascertain in practice.
+
+See `target_subset_preimage_target` for a version stated using preimages instead of images.
+-/
+lemma map_target_subset_target (h : IsImmersionAtOfComplement F I J n f x) :
+    (h.equiv ∘ (·, 0)) '' (h.domChart.extend I).target ⊆ (h.codChart.extend J).target := by
+  rw [← h.writtenInCharts.image_eq, Set.image_comp, Set.image_comp,
+    PartialEquiv.symm_image_target_eq_source, OpenPartialHomeomorph.extend_source,
+    ← PartialEquiv.image_source_eq_target]
+  have : f '' h.domChart.source ⊆ h.codChart.source := by
+    simp [h.source_subset_preimage_source]
+  grw [this, OpenPartialHomeomorph.extend_source]
+
+/-- If `f` is an immersion at `x`, its domain chart's target `(h.domChart.extend I).target`
+is mapped to it codomain chart's target `(h.domChart.extend J).target`:
+see `map_target_subset_target` for a version stated using images. -/
+lemma target_subset_preimage_target (h : IsImmersionAtOfComplement F I J n f x) :
+    (h.domChart.extend I).target ⊆ (h.equiv ∘ (·, 0)) ⁻¹' (h.codChart.extend J).target :=
+  fun _x hx ↦ h.map_target_subset_target (mem_image_of_mem _ hx)
+
+/-- If `f` is an immersion at `x` and `g = f` on some neighbourhood of `x`,
+then `g` is an immersion at `x`. -/
+lemma congr_of_eventuallyEq (hf : IsImmersionAtOfComplement F I J n f x) (hfg : f =ᶠ[𝓝 x] g) :
+    IsImmersionAtOfComplement F I J n g x := by
+  rw [IsImmersionAtOfComplement_def]
+  exact LiftSourceTargetPropertyAt.congr_of_eventuallyEq
+    isLocalSourceTargetProperty_immersionAtProp hf.property hfg
+
+/-- If `f = g` on some neighbourhood of `x`,
+then `f` is an immersion at `x` if and only if `g` is an immersion at `x`. -/
+lemma congr_iff_of_eventuallyEq (hfg : f =ᶠ[𝓝 x] g) :
+    IsImmersionAtOfComplement F I J n f x ↔ IsImmersionAtOfComplement F I J n g x := by
+  simpa only [IsImmersionAtOfComplement_def] using
+    LiftSourceTargetPropertyAt.congr_iff_of_eventuallyEq
+      isLocalSourceTargetProperty_immersionAtProp hfg
+
+lemma small (hf : IsImmersionAtOfComplement F I J n f x) : Small.{u} F :=
+  small_of_injective <| hf.equiv.injective.comp (Prod.mk_right_injective 0)
+
+/-- Given an immersion `f` at `x`, this is a choice of complement which lives in the same universe
+as the model space for the co-domain of `f`: this is useful to avoid universe restrictions. -/
+def smallComplement (hf : IsImmersionAtOfComplement F I J n f x) : Type u :=
+  have := hf.small
```

Comment:
```suggestion
  haveI := hf.small
```
(You can have a look at the generated term, it is nicer like that). Same thing below, in all the definitions and instances I'd replace `have := hf.small` with `haveI` to get nicer terms.

Reply:
I know this in principle; thanks for reminding me. I'm not sure if it's as important in the instances, but I replaced those anyway. I strongly agree on the definitions.